src/fastify-plugins/maps/index.js

```diff
@@ -20,57 +32,158 @@ export const plugin = fp(mapsPlugin, {
 /**
  * @typedef {object} MapsPluginOpts
  * @property {string} [prefix]
+ * @property {string} [defaultOnlineStyleUrl]
  */
 
 /** @type {import('fastify').FastifyPluginAsync<MapsPluginOpts>} */
 async function mapsPlugin(fastify, opts) {
   fastify.register(routes, {
     prefix: opts.prefix,
+    defaultOnlineStyleUrl: opts.defaultOnlineStyleUrl,
   })
 }
 
+const GetStyleJsonQueryStringSchema = T.Object({
+  key: T.Optional(T.String()),
+})
+
 /** @type {import('fastify').FastifyPluginAsync<MapsPluginOpts, import('fastify').RawServerDefault, import('@fastify/type-provider-typebox').TypeBoxTypeProvider>} */
-async function routes(fastify) {
-  fastify.get('/style.json', async (req, rep) => {
-    const serverAddress = await getFastifyServerAddress(req.server.server)
-
-    // 1. Attempt to get "default" local static map's style.json
-    {
-      const styleId = 'default'
-
-      const results = await Promise.all([
-        fastify.mapeoStaticMaps.getStyleJsonStats(styleId),
-        fastify.mapeoStaticMaps.getResolvedStyleJson(styleId, serverAddress),
-      ]).catch(() => {
-        fastify.log.warn('Cannot read default static map')
-        return null
-      })
-
-      if (results) {
+async function routes(fastify, opts) {
+  const { defaultOnlineStyleUrl = DEFAULT_MAPBOX_STYLE_URL } = opts
+
+  fastify.get(
+    '/style.json',
+    { schema: { querystring: GetStyleJsonQueryStringSchema } },
+    async (req, rep) => {
+      const serverAddress = await getFastifyServerAddress(req.server.server)
+
+      // 1. Attempt to get "default" local static map's style.json
+      {
+        const styleId = 'default'
+
+        const results = await Promise.all([
+          fastify.mapeoStaticMaps.getStyleJsonStats(styleId),
+          fastify.mapeoStaticMaps.getResolvedStyleJson(styleId, serverAddress),
+        ]).catch(() => {
+          fastify.log.warn('Cannot read default static map')
+          return null
+        })
+
+        if (results) {
+          const [stats, styleJson] = results
+          rep.headers(createStyleJsonResponseHeaders(stats.mtime))
+          return styleJson
+        }
+      }
+
+      // 2. Attempt to get a default style.json from online source
+      {
+        const { key } = req.query
+
+        const upstreamUrlObj = new URL(defaultOnlineStyleUrl)
+        const { hostname } = upstreamUrlObj
+
+        if (key) {
+          if (isRecognizedMapProvider(hostname)) {
+            const paramToUpsert = getMapProviderApiKeyQueryParamName(hostname)
```

Comment:
Currently, the flow is:

1. Determine if the map provider is in a list with `isRecognizedMapProvider`.
1. If it is, pass it to `getMapProviderApiKeyQueryParamName` (which could theoretically throw an error) and replace the query string.
1. If it's not, log a warning.

I wonder if we could avoid error states completely by changing the flow to:

1. Look up `paramToUpsert` in a Map.
1. If it's found, replace it.
1. If it's not, log a warning.

This solution (which might be worse for some reason) is ~35 lines shorter:

```diff
diff --git a/src/fastify-plugins/maps/index.js b/src/fastify-plugins/maps/index.js
index 6fc4cfb..9e48e67 100644
--- a/src/fastify-plugins/maps/index.js
+++ b/src/fastify-plugins/maps/index.js
@@ -14,12 +14,17 @@ export const PLUGIN_NAME = 'mapeo-maps'
 export const DEFAULT_MAPBOX_STYLE_URL =
   'https://api.mapbox.com/styles/v1/mapbox/outdoors-v12'
 
-const RECOGNIZED_MAP_PROVIDERS = /** @type {const} */ ([
-  'api.mapbox.com',
-  'api.protomaps.com',
-  'api.maptiler.com',
-  'tiles.stadiamaps.com',
-  'basemapstyles-api.arcgis.com',
+const MAP_PROVIDER_API_KEY_QUERY_PARAM_BY_HOSTNAME = new Map([
+  // Mapbox expects `access_token`: https://docs.mapbox.com/api/maps/styles/
+  ['api.mapbox.com', 'access_token'],
+  // Protomaps expects `key` (no docs link yet)
+  ['api.protomaps.com', 'key'],
+  // MapTiler expects `key`: https://docs.maptiler.com/cloud/api/maps/
+  ['api.maptiler.com', 'key'],
+  // Stadia expects `api_key`: https://docs.stadiamaps.com/themes/
+  ['tiles.stadiamaps.com', 'api_key'],
+  // ArcGIS expects `token`: https://developers.arcgis.com/documentation/mapping-apis-and-services/security/api-keys/
+  ['basemapstyles-api.arcgis.com', 'token'],
 ])
 
 export const plugin = fp(mapsPlugin, {
@@ -84,9 +89,9 @@ async function routes(fastify, opts) {
         const { hostname } = upstreamUrlObj
 
         if (key) {
-          if (isRecognizedMapProvider(hostname)) {
-            const paramToUpsert = getMapProviderApiKeyQueryParamName(hostname)
-
+          const paramToUpsert =
+            MAP_PROVIDER_API_KEY_QUERY_PARAM_BY_HOSTNAME.get(hostname)
+          if (paramToUpsert) {
             // Note that even if the search param of interest already exists in the url
             // it is overwritten by the key provided in the request's search params
             upstreamUrlObj.searchParams.set(paramToUpsert, key)
@@ -147,43 +152,3 @@ async function routes(fastify, opts) {
     }
   )
 }
-
-/**
- * @param {string} hostname URL hostname
- * @returns {hostname is typeof RECOGNIZED_MAP_PROVIDERS[number]}
- */
-function isRecognizedMapProvider(hostname) {
-  return RECOGNIZED_MAP_PROVIDERS.includes(
-    /** @type {typeof RECOGNIZED_MAP_PROVIDERS[number]} */ (hostname)
-  )
-}
-
-/**
- * @param {typeof RECOGNIZED_MAP_PROVIDERS[number]} hostname
- * @returns {string} The name of the URL search param used by a provider to specify an auth-related API key or token
- */
-function getMapProviderApiKeyQueryParamName(hostname) {
-  switch (hostname) {
-    // Mapbox expects `access_token`: https://docs.mapbox.com/api/maps/styles/
-    case 'api.mapbox.com': {
-      return 'access_token'
-    }
-    // Protomaps expects `key` (no docs link yet)
-    // MapTiler expects `key`: https://docs.maptiler.com/cloud/api/maps/
-    case 'api.protomaps.com':
-    case 'api.maptiler.com': {
-      return 'key'
-    }
-    // Stadia expects `api_key`: https://docs.stadiamaps.com/themes/
-    case 'tiles.stadiamaps.com': {
-      return 'api_key'
-    }
-    // ArcGIS expects `token`: https://developers.arcgis.com/documentation/mapping-apis-and-services/security/api-keys/
-    case 'basemapstyles-api.arcgis.com': {
-      return 'token'
-    }
-    default: {
-      throw new Error(`Invalid value provided ${hostname}`)
-    }
-  }
-}
```

Take it or leave it!

Reply:
good suggestion! makes way more sense looking back on it :)

Reply:
addressed via https://github.com/digidem/mapeo-core-next/pull/459/commits/c0c25d9fa5204f76a5d04c026da0b039368afe58